# 异步函数

## await

因为异步函数主要针对不会发上完成的任务，所以自然需要一种暂停和恢复执行的能力，使用await 关键字可以暂停异步函数代码的执行，等待期约的解决。

**以下两种写法是相同的**

Promise

```js
let p = new Promise((resolve, reject)=>setTimeout(console.log, 1000, 3))
p.then(console.log)
```


异步函数

```js
async function foo(params) {
  let p = new Promise((resolve, reject)=>setTimeout(console.log, 1000, 3))
  console.log(await p);
}
foo()
```


> await 关键字的原理: await 关键字会暂停执行异步函数后面的代码，让出javascript 的执行线程。这个行为与生成器函数中的 yield 关键字是一样的。await 关键字同样是尝试解包对象的值，然后将这个值传给表达式， 再异步恢复异步函数的执行。


## await与异步函数的返回

await 期待(实际上并不要求)一个实现thenable接口的对象，但常规的值也可以。如果是一个实现thenable接口的对象，则这个对象会被await 默认解包; 如果不是，则这个值会被当作已经解决的期约。



1. 等待一个原始值

```js
const foo = async () => {
  const a = await "foa";
  console.log("a", a);
};
foo();
// a foa
```


2. **等待一个实现了thenable接口的对象**

```js
const foo1 = async () => {
  const thenable = {
    then(callback) {
      callback("baz");
    },
  };
  const a = await thenable;
  console.log("a", a);
};

foo1();
// a baz
```


3. 等待一个期约

```js
const foo = async () => {
  const a = await Promise.resolve("hello");
  console.log("a", a);
};

foo();
// a hello
```



4. 等待会抛出错误的同步操作，会返回拒绝的期约, 抛出未捕获错误

```js
const foo = async () => {
  console.log(1);
  await (() => {
    throw "3";
  })();
};

foo()
console.log(2);
// 1
// 2
// (node:8597) UnhandledPromiseRejectionWarning: 3
```

```js
const foo = async () => {
  console.log(1);
  await (() => {
    throw "3";
  })();
};

foo().catch(console.log);
console.log(2);
// 1
// 2
// 3
```


5. reject 后面的代码不会执行

```js
const foo = async () => {
  console.log(1);
  await Promise.reject(3);
  console.log(4) // 这里不会执行
};

foo().catch(console.log);
console.log(2);

```


## await理解

> 要完全理解await 关键字，要知道它并非仅仅是等待一个值那么简单，javascript 在碰到 await关键字时，会记录在哪里暂停执行，等到await 右侧的值可用了，javascript运行时会向消息队列推送一个任务，这个任务会恢复异步任务的执行。
> 因此，即使await 后面跟着一个立即可用的值，函数的其余部分也会异步求值。


看以下例子

```js
async function b() {
  console.log(1);
  await null;
  console.log(2);
}

b();
console.log(3);
// 1
// 3
// 2
```


如果await等待的是一个契约


```js
async function b() {
  console.log(1);
  console.log(await 2);
  console.log(3);
}

b();
console.log(4);
// 1
// 4
// 2
// 3

```

下面的打印结果同上

```js
async function b() {
  console.log(1);
  console.log(await Promise.resolve(2));
  console.log(3);
}

b();
console.log(4);
// 1
// 4
// 2
// 3
```



## 异步函数策略

### 实现sleep


实现非阻塞的暂停

```js
function sleep(delay) {
  return new Promise((resolve) => setTimeout(resolve, delay));
}

async function foo() {
  const t0 = Date.now();
  await sleep(1500);
  console.log(Date.now() - t0);
}
foo();
// 1501
```




<!-- ### 实现串行 -->
## 练习题

以下代码的执行顺序是什么？

1. 考察是否知道await后代码的执行顺序


```js
function foo(){
  awiat console.log(1);
}
foo()
console.log(2)
```


2. 考察是否知道Promise.reject后的代码不执行

```js
const foo = async () => {
  console.log(1);
  await Promise.reject(3);
  console.log(4) // 这里不会执行
};

foo().catch(console.log);
console.log(2);
```